crates/supervisor/storage/src/chaindb.rs

```diff
@@ -284,6 +284,67 @@ impl HeadRefStorageWriter for ChainDb {
         })?
     }
 
+    fn update_current_cross_unsafe(&self, block: &BlockInfo) -> Result<(), StorageError> {
+        self.observe_call("update_current_cross_unsafe", || {
+            self.env.update(|tx| {
+                let lp = LogProvider::new(tx);
+                let sp = SafetyHeadRefProvider::new(tx);
+
+                // Check parent-child relationship with current CrossUnsafe head, if it exists.
+                let parent = sp.get_safety_head_ref(SafetyLevel::CrossUnsafe)?;
+                if !parent.is_parent_of(block) {
+                    return Err(StorageError::ConflictError(
+                        "candidate block is not the child of the current cross-unsafe head"
+                            .to_string(),
+                    ));
+                }
+
+                // Ensure the block exists in log storage and hasn't been pruned due to a re-org.
+                let stored_block = lp.get_block(block.number)?;
+                if stored_block.hash != block.hash {
+                    warn!(
+                        target: "supervisor_storage",
+                        incoming_block_hash = %block.hash,
+                        stored_block_hash = %stored_block.hash,
+                        "Hash mismatch while updating CrossUnsafe head",
+                    );
+                    return Err(StorageError::EntryNotFound(
+                        "block hash does not match".to_string(),
+                    ));
+                }
+
+                sp.update_safety_head_ref(SafetyLevel::CrossUnsafe, block)?;
+
+                Ok(())
+            })?
+        })
+    }
+
+    fn update_current_cross_safe(&self, block: &BlockInfo) -> Result<DerivedRefPair, StorageError> {
+        self.observe_call("update_current_cross_safe", || {
```

Comment:
Could you updated this in Metric::METHODS?

Reply:
added